packages/arb-token-bridge-ui/src/hooks/useTransactionHistory.ts

```diff
@@ -332,35 +337,138 @@ export const useTransactionHistory = (
     (tx: MergedTransaction) => {
       if (!isTxPending(tx)) {
         return
       }
 
-      mutate(currentNewTransactions => {
+      mutateNewTransactionsData(currentNewTransactions => {
         if (!currentNewTransactions) {
           return [tx]
         }
 
         return [tx, ...currentNewTransactions]
       })
     },
-    [mutate]
+    [mutateNewTransactionsData]
+  )
+
+  const updateCachedTransaction = useCallback(
+    (newTx: MergedTransaction) => {
+      // check if tx is a new transaction initiated by the user, and update it
+      const foundInNewTransactionsCache = !!newTransactionsData?.find(oldTx =>
+        isSameTransaction(oldTx, newTx)
+      )
+
+      if (foundInNewTransactionsCache) {
+        // replace the existing tx with the new tx
+        mutateNewTransactionsData(txs =>
+          txs?.map(oldTx => {
+            return { ...(isSameTransaction(oldTx, newTx) ? newTx : oldTx) }
+          })
+        )
+        return
+      }
+
+      // tx not found in the new user initiated transaction list
+      // look in the paginated historical data
+      if (!txPages) {
+        return
+      }
+
+      let pageNumberToUpdate = 0
+
+      // search cache for the tx to update
+      while (
+        !txPages[pageNumberToUpdate]?.find(oldTx =>
+          isSameTransaction(oldTx, newTx)
+        )
+      ) {
+        pageNumberToUpdate++
+
+        if (pageNumberToUpdate > txPages.length) {
+          // tx not found
+          return
+        }
+      }
+
+      const oldPageToUpdate = txPages[pageNumberToUpdate]
+
+      if (!oldPageToUpdate) {
+        return
+      }
+
+      // replace the old tx with the new tx
+      const updatedPage = oldPageToUpdate.map(oldTx => {
+        return isSameTransaction(oldTx, newTx) ? newTx : oldTx
+      })
+
+      // all old pages including the new updated page
+      const newTxPages = [
+        ...[...txPages].slice(0, pageNumberToUpdate),
+        updatedPage,
+        ...[...txPages].slice(pageNumberToUpdate + 1)
+      ]
+
+      mutateTxPages(newTxPages, false)
+    },
+    [mutateNewTransactionsData, mutateTxPages, newTransactionsData, txPages]
+  )
+
+  const updatePendingTransaction = useCallback(
+    async (tx: MergedTransaction) => {
+      const foundInCache = !!transactions.find(t => isSameTransaction(t, tx))
+
+      if (!foundInCache) {
+        // tx does not exist
+        return
+      }
+
+      if (!isTxPending(tx)) {
+        // if not pending we don't need to check for status, we accept whatever status is passed in
+        updateCachedTransaction(tx)
+        return
+      }
+
+      if (tx.isCctp) {
+        const updatedCctpTransfer = await getUpdatedCctpTransfer(tx)
+        updateCachedTransaction(updatedCctpTransfer)
+        return
+      }
+
+      if (tx.isWithdrawal) {
+        const updatedWithdrawal = await getUpdatedWithdrawal(tx)
+        updateCachedTransaction(updatedWithdrawal)
+      } else {
```

Comment:
Instead of if/else let's early return in the withdrawal case, and do the same with the deposits instead of the ternary. Should be easier to read

Reply:
updated